repl/src/main/scala/org/apache/spark/repl/Signaling.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.repl
+
+import org.apache.spark.SparkContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{Signaling => USignaling}
+
+private[repl] object Signaling extends Logging {
+
+  /**
+   * Register a SIGINT handler, that terminates all active spark jobs or terminates
+   * when no jobs are currently running.
+   * This makes it possible to interrupt a running shell job by pressing Ctrl+C.
+   */
+  def cancelOnInterrupt(ctx: SparkContext): Unit = USignaling.register("INT") {
+    if (!ctx.statusTracker.getActiveJobIds().isEmpty) {
+      logWarning("Cancelling all active jobs, this can take a while. " +
+        "Press Ctrl+C again to exit now.")
```

Comment:
Oh, I think I missunderstood your comment, are you proposing of dropping
the in-job condition and always catch a C-c?
In that case I'm not sure that I agree, since the main purpose of a sigint
is to interrupt a process. Killing jobs first and catching the signal is a
compromise for convenience. Imagine a spark shell that is stuck in some
arbitrary user code (not running a spark job), in that case C-c should imo
still exit the process.
On Apr 21, 2016 10:51 AM, "Jakob Odersky" <jakob@odersky.com> wrote:

> C-d is slightly different in that it just emits an end-of-transmission
> (eot) character to the active terminal.
> When the shell is running a job, it cannot listen to input and hence C-d
> has no effect.
>
> C-c on the other hand, sends an asynchronous signal by the host operating
> system, which will be executed by the signal handler in a separate thread
> and thus be able to stop running jobs.
> On Apr 21, 2016 10:13 AM, "Davies Liu" <notifications@github.com> wrote:
>
> In repl/src/main/scala/org/apache/spark/repl/Signaling.scala
> <https://github.com/apache/spark/pull/12557#discussion_r60619566>:
>
> > +
> > +import org.apache.spark.SparkContext
> > +import org.apache.spark.internal.Logging
> > +import org.apache.spark.util.{Signaling => USignaling}
> > +
> > +private[repl] object Signaling extends Logging {
> > +
> > +  /**
> > +   * Register a SIGINT handler, that terminates all active spark jobs or terminates
> > +   * when no jobs are currently running.
> > +   * This makes it possible to interrupt a running shell job by pressing Ctrl+C.
> > +   */
> > +  def cancelOnInterrupt(ctx: SparkContext): Unit = USignaling.register("INT") {
> > +    if (!ctx.statusTracker.
> getActiveJobIds().isEmpty) {
> > +      logWarning("Cancelling all active jobs, this can take a while. " +
> > +        "Press Ctrl+C again to exit now.")
>
> We can type Ctrl+D to quit the shell, should we always use Ctrl+C to
> cancel the running job and clear the current line input?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/12557/files/4f9bf695344a5c4c54372eaa6bf54af0d2da1f74#r60619566>
>
>